templates/widget/__tests__/widget.test.ts

```diff
@@ -0,0 +1,33 @@
+import 'reflect-metadata';
+import { MessageClient, MessageService } from '@theia/core';
+import { ContainerModule, Container } from 'inversify';
+import { MockMessageClient, MockMessageService } from './mock-objects/mock-message-service';
+import { <%= params.extensionPrefix %>Widget } from '../<%= params.extensionPath %>-widget';
+import { render } from '@testing-library/react'
+
+describe('widget extension unit tests', () => {
+
+    let widget: any;
+
+    beforeEach(async () => {
+        const module = new ContainerModule( bind => {
+            bind(MessageClient).to(MockMessageClient).inSingletonScope();
+            bind(MessageService).to(MockMessageService).inSingletonScope();
+            bind(<%= params.extensionPrefix %>Widget).toSelf();
+        });
+        const container = new Container();
+        container.load(module);
+        widget = container.resolve<<%= params.extensionPrefix %>Widget>(<%= params.extensionPrefix %>Widget);
+    });
+
+    it('should render react node correctly', async () => {
+        const element = render(widget.render());
+        expect(element.queryByText('Display Message')).toBeTruthy();
+    });
+
+    it('should inject the message service', async () => {
+        widget.displayMessage();
+        expect(widget.messageService.infoWasCalled).toBe(true);
+    });
```

Comment:
This test is likely overkill for what it actually tests.

We can do it the following way (without the need to mock the service):

```ts
describe('widget extension unit tests', () => {

    let widget: any;

    beforeEach(async () => {
        const module = new ContainerModule(bind => {
            bind(MessageClient).to(MockMessageClient).inSingletonScope();
            bind(MessageService).toConstantValue({
                info(message: string): void {
                    console.log(message);
                }
            } as MessageService);
            bind(CustomWidgetWidget).toSelf();
        });
        const container = new Container();
        container.load(module);
        widget = container.resolve<CustomWidgetWidget>(CustomWidgetWidget);
    });

    it('should be an instance of \'CustomWidgetWidget\'', () => {
        expect(widget instanceof CustomWidgetWidget).toBeTruthy();
    });

    it('should render react node correctly', async () => {
        const element = render(widget.render());
        expect(element.queryByText('Display Message')).toBeTruthy();
    });

    it('should inject \'MessageService\'', () => {
        const spy = jest.spyOn(widget, 'displayMessage')
        widget.displayMessage();
        expect(spy).toBeCalled();
    });

});
```

Reply:
Yes, we were discussing the trade off between adding a "spyOn" as an additional concept an potential complexity vs. a dedicated Mock. However, looking at your proposal, I am convinced that this is more comprehensible. I adapted it.